### PR TITLE
Add support for nan-euclidean distance measure

### PR DESCRIPTION
**Pull request by dzlab**
Add support for the nan-euclidean distance measure as available in sklearn's [KNNImputer](https://scikit-learn.org/stable/modules/generated/sklearn.impute.KNNImputer.html)

Changes:
- Added new param `metric` for specifying the distance measure metric
- Extracted the logic that calculates distance between `VectorWithNorm` into a trait
- Added `EuclideanDistanceMetric` as default measure
- Added `NaNEuclideanDistanceMetric` as the implementation for the nan-euclidean distance measure

### CODE REVIEW

**Comment by dzlab** [comment]
@saurfang can you review this PR, thanks

**Review by saurfang — APPROVED** [review]
awesome. thanks for the contribution
